### ISSUE TITLE
Plugin System

### ISSUE THREAD

**Issue by KilledByAPixel** [KilledByAPixel/LittleJS#86]
It would be good to have a plugin system for littles where other components can live that are separate from the main engine.

The medals system, or at least the newgrounds implementation, would be moved to being a plugin.

Other ideas for plugins...
- Box2d integration
- 3d rendering system
- 2d dynamic lighting system
- post processing effects
- procedural texture generation
- AI system
- multiplayer

Still thinking about the best way to set it up. What plugins would you like to see?

**Comment by rishavs** [comment]
camera
Steam Integration
joystick integration

**Comment by KilledByAPixel** [comment]
We have joystick integration already, steam would be amazing though!

- ui framework
- level editor
- animation system
- pathfinding
- import from popular level editors like Tiled

**Comment by thewrath** [comment]
If you give me some guidelines (on the expected result) I can try to work on it.

I could, for example, suggest that i try to re-implement the medal system as a plugin and thus make the necessary extension points appear. 🙂

**Comment by KilledByAPixel** [comment]
Thank you, but the main work here is really planning out how to set up the plugin system. I have a rough idea and hope to get to it soon. The medals system is already totally separate from everything else so it should be simple to move it to a plugin

**Comment by KilledByAPixel** [comment]
Spine 2d animation tool would be good for a plugin...

[Spine Documentation](https://esotericsoftware.com/spine-documentation#Spine-Documentation)

Also I just added Tiled editor support to the platformer demo, that could also be moved to a plugin eventually.

**Comment by KilledByAPixel** [comment]
We could integrate audio convolvers pretty easily as a plugin. This is a good demo...

https://mdn.github.io/voice-change-o-matic/

**Comment by KilledByAPixel** [comment]
We now how a plugin system! It is pretty simple, plugins can register for render or update callbacks. We can add more if necessary. I moved newgrounds and post processing to plugins. Also I created Box2D as a plugin.

**Comment by hockdudu** [comment]
Is it possible for the plugin system to also support ESM integration?

**Comment by KilledByAPixel** [comment]
> Is it possible for the plugin system to also support ESM integration?

Yes, that is a goal I hope to do soon! I just need to add exports and stuff, the first step was just getting the plugin system started.

**Comment by KilledByAPixel** [comment]
a 9 slice object where you can split up a sprite 9 ways to stretch it better.

https://youtu.be/37q-0oS65LY?si=CUDAtTurAVQraxni

**Comment by KilledByAPixel** [comment]
I added more plugins...

Tween system: https://killedbyapixel.github.io/LittleJS/examples/tweenSystem
Pathfinding: https://killedbyapixel.github.io/LittleJS/examples/?example=Path+Finder

Also I moved the medals to plugins since they are already insulated from the rest of the engine. This keeps the core engine more clean.

**Comment by KilledByAPixel** [comment]
Many of the plugins mentioned earlier are now in the engine. Here is list of most wanted plugins...

- animation system
- 3d rendering system
- 2d dynamic lighting system
- improved post processing tools
- level editor
- multiplayer
